airbyte-integrations/connectors/source-google-ad-manager/source_google_ad_manager/streams.py

```diff
@@ -0,0 +1,324 @@
+from typing import Any, Mapping, Union, List
+from csv import DictReader as csv_dict_reader
+from .data_classes import AdUnitPerHourItem, AdUnitPerReferrerItem
+
+_CHUNK_SIZE = 16 * 1024
+
+
+logger = logging.getLogger('{}.{}'.format(__name__, 'google_ad_manager_report_downloader'))
+
+
+class BaseGoogleAdManagerReportStream(Stream, ABC):
+    """
+    this is the base stream class used to generate the report
+    """
+
+    def __init__(self, google_ad_manager_client: ad_manager.AdManagerClient) -> None:
+        super().__init__()
+        self.google_ad_manager_client = google_ad_manager_client
+        self.report_downloader = self.google_ad_manager_client.GetDataDownloader(version='v202208')
+    
+    def download_report(self, report_job_id: str, export_format: str, include_report_properties: bool = False,
+                        include_totals_row: bool = None, use_gzip_compression: bool = True) -> requests.Response:
+        """make an api call to the api with the report job id and return a response containing the report
+
+        Args:
+          report_job_id: The ID of the report job to wait for, as a string.
+          export_format: The export format for the report file, as a string.
+          outfile: A writeable, file-like object to write to.
+          include_report_properties: Whether or not to include the report
+            properties (e.g. network, user, date generated...)
+            in the generated report.
+          include_totals_row: Whether or not to include the totals row.
+          use_gzip_compression: Whether or not to use gzip compression.
+        """
+        service = self.report_downloader._GetReportService()
+
+        if include_totals_row is None:  # True unless CSV export if not specified
+            include_totals_row = True if export_format != 'CSV_DUMP' else False
+        opts = {
+            'exportFormat': export_format,
+            'includeReportProperties': include_report_properties,
+            'includeTotalsRow': include_totals_row,
+            'useGzipCompression': use_gzip_compression
+        }
+        report_url = service.getReportDownloadUrlWithOptions(report_job_id, opts)
+        logger.info('Request Summary: Report job ID: %s, %s', report_job_id, opts)
+
+        response = self.report_downloader.url_opener.open(report_url)
+        return response
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        """
+        this is the base method to parse the api response
+        """
+        logger.info("start parsing the response stream, should be replaced with logger")
+
+        while True:
+            chunk = response.read(_CHUNK_SIZE)
+            if not chunk:
+                break
+            lines = chunk.decode('utf-8')
+            reader = csv_dict_reader(lines.splitlines())
+            for row in reader:
+                item = self.generate_item(row)
+                yield item.dict()
```

Comment:
You need to specify a timeout to not have an infinity loop here.

Reply:
👍🏻